### DISCUSSION TITLE
p2artifacts.xml, how is that generated? By default it doesn't have certain properties

### DISCUSSION THREAD

**Discussion by jcompagner** [eclipse-tycho/tycho#1549] [General]
I think that is be done by the tycho-p2-repository plugin and then i guess the p2-metadata section.
But a lot of docs are not there anymore (some links for example in the release notes of Tycho 3.0 are not working because of this stuff like: https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/ )

now if i build a plugin it generates:

```
<?xml version='1.0' encoding='UTF-8'?>
<?artifactRepository version='1.1.0'?>
<artifacts size='1'>
  <artifact classifier='osgi.bundle' id='org.eclipse.jface' version='3.27.0.v20220817-1444'>
    <properties size='6'>
      <property name='artifact.size' value='1187679'/>
      <property name='download.size' value='1187679'/>
      <property name='maven-groupId' value='eclipse.platform.ui'/>
      <property name='maven-artifactId' value='org.eclipse.jface'/>
      <property name='maven-version' value='3.27.0.v20220817-1444'/>
      <property name='maven-type' value='eclipse-plugin'/>
    </properties>
  </artifact>
</artifacts>
```

but what tycho wants is the .m2\repository\p2\osgi\bundle\org.eclipse.jface\3.27.0.v20220817-1444
location is:

```
<?xml version='1.0' encoding='UTF-8'?>
<?artifactRepository version='1.1.0'?>
<artifacts size='1'>
  <artifact classifier='osgi.bundle' id='org.eclipse.jface' version='3.27.0.v20220817-1444'>
    <properties size='10'>
      <property name='artifact.size' value='1168114'/>
      <property name='download.size' value='1168114'/>
      <property name='maven-groupId' value='eclipse.platform.ui'/>
      <property name='maven-artifactId' value='org.eclipse.jface'/>
      <property name='maven-version' value='3.27.0-SNAPSHOT'/>
      <property name='maven-type' value='eclipse-plugin'/>
      <property name='download.stats' value='org.eclipse.jface/3.27.0.v20220817-1444'/>
      <property name='download.md5' value='99cdd01e4b09db72c2e28e4c2efccd12'/>
      <property name='download.checksum.md5' value='99cdd01e4b09db72c2e28e4c2efccd12'/>
      <property name='download.checksum.sha-256' value='ebab5be009308fb68f694ea521bb95b199ff43d36a30fc3549783d464180474a'/>
    </properties>
  </artifact>
</artifacts>

```

the thing is what i do is 
https://github.com/Servoy/servoy-eclipse/blob/master/org.eclipse.jface/pom.xml#L60

because we have a patch jface plugin (that as far as i see we really need because we need to be able to override all icons that our product has no matter which plugin they come from, so there is 1 extra class and 1 class is adjusted for this)

until now (tycho 2.6.0) this worked fine.
But now with 2.7.5 it always starts to download when it seems my generated p2artifacts.xml:

```
[WARNING] The following locally built units have been used to resolve project dependencies:
[WARNING]   servoy_ngclient/2022.12.0.3820_rc
[WARNING]   org.eclipse.dltk.javascript.rhino/1.7.14.s1
[WARNING]   com.servoy.eclipse.nodejs.linux.gtk.x86_64/16.17.0
[WARNING]   org.apache.tomcat/9.0.50
[WARNING]   org.eclipse.jface/3.27.0.v20220817-1444
[WARNING]   j2db_log4j/2022.12.0.3820_rc
[WARNING]   j2db_server/2022.12.0.3820_rc
[WARNING]   com.servoy.eclipse.ngclient/2022.12.0.3820_rc
[WARNING]   servoy_smart_client/2022.12.0.3820_rc
[WARNING]   com.servoy.eclipse.model/2022.12.0.3820_rc
[WARNING]   sablo/2022.12.0.3820_rc
[WARNING]   servoy_debug/2022.12.0.3820_rc
[WARNING]   com.servoy.eclipse.ngclient.ui/2022.12.0.3820_rc
[WARNING]   servoy_headless_client/2022.12.0.3820_rc
[WARNING]   servoy_shared/2022.12.0.3820_rc
[WARNING]   servoy_base/2022.12.0.3820_rc
[INFO] Fetching 202209141001&countryCode=us&timeZone=1&format=xml from https://www.eclipse.org/downloads/download.php?format=xml&file=/releases/2022-09/
[INFO] Fetching org.eclipse.jface_3.27.0.v20220817-1444.jar from https://mirror.aarnet.edu.au/pub/eclipse/releases/2022-09/202209141001/plugins/ (1,11MB)
[INFO] Fetching org.eclipse.jface_3.27.0.v20220817-1444.jar from https://mirror.aarnet.edu.au/pub/eclipse/releases/2022-09/202209141001/plugins/ (210,73kB of 1,11MB at 206,73kB/s)
```

so at the first time it encounters the jface plugin, Tycho doesn't find the one in my local repo to be correct and will download the original one of eclipse..

So it seems to me that i really need to generate such a full artifacts.xml file. Is that configurable?

It seems if i don't overwrite the xml file that Tycho finds it ok.. But that is then with invalid data (sizes and checksums)

**Comment by laeubi** [comment]
Using **the same version and id** is really bad and because of that Tycho detects that your jar is tampered and thus triggers a fresh download.

So you better ask at jface to provide some kind of extension point for what you want to archive here, or choose a higher version for your patch e.g. `3.27.0.z-patches-v20220817-1444`

**Reply by jcompagner** [reply]
no, this is already a long long discussion for years (in bugs.eclipse.org), we tried all kind of stuff (including class weaving). But that is just a no go

I need to patch the actual jar or be able to use a Fragment (patch). But i can't get that to work (my other projects using the new classes just don't want to see this)

We need to be able to patch it, so how can i tell Tycho (just as in 2.6) to use my file?

We really need to patch exactly that jface version, because that is what the core eclipse plugins are using and we need our patch running directly.

This current patching is already working for years now. So this just has to work. Our design people really want to provide for our product all kinds of different icons (we replace pretty much all icons of all the 3rd party plugins that we bundle in our product)

**Reply by laeubi** [reply]
Patch features are something that is barely supported by Tycho at all as no of the committers or platform regulatory using it. I even doubt there is not any tests for it. So if patch-features are crucial for you, I think the best start would be to provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue.

Beside that, if it is crucial to your business you can find dedicated support options for Tycho here:
https://github.com/eclipse-tycho/tycho#getting-support

**Reply by jcompagner** [reply]
using a different version doesn't work:

```
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: com.servoy.eclipse.feature.feature.group 2022.12.0.3820_rc
[ERROR]   Only one of the following can be installed at once: [org.eclipse.jface 3.26.0.v20220513-0449, org.eclipse.jface 3.24.0.v20211110-1517, org.eclipse.jface 3.25.0.v20220218-1636, org.eclipse.jface 3.21.0.v20200821-1458, org.eclipse.jface 3.27.0.v20220817-1444, org.eclipse.jface 3.27.0.patch-v20220817-1444]
[ERROR]   Cannot satisfy dependency: com.servoy.eclipse.feature.feature.group 2022.12.0.3820_rc depends on: org.eclipse.equinox.p2.iu; org.eclipse.jface [3.27.0.patch-v20220817-1444,3.27.0.patch-v20220817-1444]
[ERROR]   Cannot satisfy dependency: com.servoy.eclipse.feature.feature.group 2022.12.0.3820_rc depends on: org.eclipse.equinox.p2.iu; org.eclipse.rcp.feature.group 3.5.0.0
[ERROR]   Cannot satisfy dependency: org.eclipse.e4.rcp.feature.group 4.25.0.v20220829-1402 depends on: org.eclipse.equinox.p2.iu; org.eclipse.jface [3.27.0.v20220817-1444,3.27.0.v20220817-1444]
[ERROR]   Cannot satisfy dependency: org.eclipse.rcp.feature.group 4.25.0.v20220831-1800 depends on: org.eclipse.equinox.p2.iu; org.eclipse.e4.rcp.feature.group [4.25.0.v20220829-1402,4.25.0.v20220829-1402]
```


because yes the base e4.rcp feature is hardcoded to that jface version. So we can't introduce a new one.

i am not directly talking about Patch Features but more fragment projects:

https://wiki.eclipse.org/Steps_to_use_Fragments_to_patch_a_plug-in

But even without tycho, i can't get that to work inside my eclipse.... because the plugin that depends on this patch (that has a package import of org.eclipse.jface.resources) doesn't see this coming from that Fragment..
(could be that i do something wrong but i can't tell what that is currently)

so there is no way to tell the:
--- tycho-p2-plugin:2.7.5:p2-metadata-default (default-p2-metadata-default) @ org.eclipse.jface ---

to generate those other properties?

**Reply by laeubi** [reply]
> i am not directly talking about Patch Features but more fragment projects:
> https://wiki.eclipse.org/Steps_to_use_Fragments_to_patch_a_plug-in

I think the same applies here ...

> so there is no way to tell the: --- tycho-p2-plugin:2.7.5:p2-metadata-default (default-p2-metadata-default) @ org.eclipse.jface ---
> to generate those other properties?

No and as explained I don't think this can really work with current Tycho, because you can have projects that want to use the "patched" variant and other who want to use the original so this obviously would interfere if everything is stored as the same location. So it has to work in a way that such "patches" are stored in a special (sub)folder with additional data that is sued in such a case while keeping the original data from the updatesite intact.

**Reply by jcompagner** [reply]
hmm if i just don't override the p2artifacts.xml in that p2/osgi area
but just the jar.. then the product is fine (it has my patched jar)
but i was then very afraid that the repository would have the wrong data... (the artifacts.xml)

but if i check that..

```
    <artifact classifier='osgi.bundle' id='org.eclipse.jface' version='3.27.0.v20220817-1444'>
      <properties size='10'>
        <property name='artifact.size' value='1187658'/>
        <property name='download.size' value='1187658'/>
        <property name='maven-groupId' value='eclipse.platform.ui'/>
        <property name='maven-artifactId' value='org.eclipse.jface'/>
        <property name='maven-version' value='3.27.0-SNAPSHOT'/>
        <property name='maven-type' value='eclipse-plugin'/>
        <property name='download.stats' value='org.eclipse.jface/3.27.0.v20220817-1444'/>
        <property name='download.md5' value='d867812d01cb01e7c8f06044a5c51ec8'/>
        <property name='download.checksum.md5' value='d867812d01cb01e7c8f06044a5c51ec8'/>
        <property name='download.checksum.sha-256' value='eb85db9a4d4e596bcc474d6b363fcbda99de25d3b62c9063fb43bd83696e6ba6'/>
      </properties>
    </artifact>
```

that is not a copy of:

```
<artifact classifier='osgi.bundle' id='org.eclipse.jface' version='3.27.0.v20220817-1444'>
    <properties size='10'>
      <property name='artifact.size' value='1168114'/>
      <property name='download.size' value='1168114'/>
      <property name='maven-groupId' value='eclipse.platform.ui'/>
      <property name='maven-artifactId' value='org.eclipse.jface'/>
      <property name='maven-version' value='3.27.0-SNAPSHOT'/>
      <property name='maven-type' value='eclipse-plugin'/>
      <property name='download.stats' value='org.eclipse.jface/3.27.0.v20220817-1444'/>
      <property name='download.md5' value='99cdd01e4b09db72c2e28e4c2efccd12'/>
      <property name='download.checksum.md5' value='99cdd01e4b09db72c2e28e4c2efccd12'/>
      <property name='download.checksum.sha-256' value='ebab5be009308fb68f694ea521bb95b199ff43d36a30fc3549783d464180474a'/>
    </properties>
  </artifact>
```

so at leat that is ok

So for now it seems fine if i just push the jar and don't touch the xml file


i don't get why you talk about 2 variants,, in the end there is just one.. that is mine. The other one should be completely gone, in the product there is only 1 jar that should be our patched one. There are no other projects that want to use the original one, the orginal one is fully replaced.

**Reply by laeubi** [reply]
> hmm if i just don't override the p2artifacts.xml in that p2/osgi area but just the jar.. then the product is fine (it has my patched jar) but i was then very afraid that the repository would have the wrong data... (the artifacts.xml)

The `artifacts.xml` is not directly copied but only used as a base and a copy of the original properties published in the source P2 repository. For a local build item, most properties do not exits though...

> So for now it seems fine if i just push the jar and don't touch the xml file

It seems fine if you neglect the fact that your patched variant is used system-wide for any builds an not only for your specific product...

> i don't get why you talk about 2 variants,, in the end there is just one.. that is mine. The other one should be completely gone, in the product there is only 1 jar that should be our patched one.
> There are no other projects that want to use the original one, the orginal one is fully replaced.

Yes in this **product** but you currently overwrite it **for every build on your machine** and that is not so "ok" ...

**Reply by jcompagner** [reply]
yes it is, i want to use that patched jface plugin to be used on every build on my local machine or on jenkins that is using that specific version of jface..
There is no other product or build that shouldn't be using it.  (for that version, of course older build or in the future new builds will have another jface plugin, but that is also patched in that version)

**Reply by mickaelistria** [reply]
> yes it is, i want to use that patched jface plugin to be used on every build on my local machine or on jenkins that is using that specific version of jface..

I don't think it's possible; and I don't think plain Maven allows it either, does it?

**Reply by laeubi** [reply]
> > yes it is, i want to use that patched jface plugin to be used on every build on my local machine or on jenkins that is using that specific version of jface..
> 
> I don't think it's possible; and I don't think plain Maven allows it either, does it?

As said before actually Tycho must "install" a patch/fragment differently (e.g.  `.m2\repository\p2\osgi\bundle\org.eclipse.jface\3.27.0.v20220817-1444\my.patch.feature\org.eclipse.jface\3.27.0.v20220817-1444.jar`) and handle consumption equally different then and then at install time use that patched jar instead.

**Reply by jcompagner** [reply]
thx for the discussion ;)

this forced me to think about it a bit more (didn't really think about this for years, because it worked until now)

But i suddenly thought, wait.. what happens if i make that patched jface plugin part of the reactor (the build) of the full product...

```
[INFO] servoy-eclipse                                                     [pom]
[INFO] org.eclipse.jface                                       [eclipse-plugin]
[INFO] com.servoy.eclipse.jre.linux.x86_64                     [eclipse-plugin]
[INFO] com.servoy.eclipse.jre.macosx.x86_64                    [eclipse-plugin]
[INFO] com.servoy.eclipse.jre.macosx.aarch64                   [eclipse-plugin]
[INFO] com.servoy.eclipse.jre.win32.x86_64                     [eclipse-plugin]
[INFO] com.servoy.eclipse.nodejs.linux.gtk.x86_64              [eclipse-plugin]
[INFO] com.servoy.eclipse.nodejs.macosx.cocoa.x86_64           [eclipse-plugin]
[INFO] com.servoy.eclipse.nodejs.macosx.cocoa.aarch64          [eclipse-plugin]
[INFO] com.servoy.eclipse.nodejs.win32.win32.x86_64            [eclipse-plugin]
[INFO] com.servoy.eclipse.feature                             [eclipse-feature]
[INFO] com.servoy.eclipse.product                          [eclipse-repository]
```

so now the .jface and the .product is in the same build..
then it also works nicely (i disabled that push to that .m2\repository\p2\osgi\bundle\org.eclipse.jface\ dir of the jar and the xml file)

The product and the repository/update site has the jar that we want.

what is a bit weird that now .m2\repository\p2\osgi\bundle\org.eclipse.jface\3.27.0.v20220817-1444\ that dir never gets an artifact. 
i guess that is not weird in our normal product. but we also have 1 other product build (which can have that patched jar just fine)

if i build that then i see:

```
[INFO] Resolving dependencies of MavenProject: com.servoy:com.servoy.eclipse.docgenerator.client.product:1.6.0 @ C:\Users\jcomp\git\servoy_master\docgenerator-ui\com.servoy.eclipse.docgenerator.client.product\pom.xml
[WARNING] The following locally built units have been used to resolve project dependencies:
[WARNING]   servoy_ngclient/2022.12.0.3820_rc
[WARNING]   com.servoy.eclipse.docgenerator.client/1.6.0
[WARNING]   org.eclipse.dltk.javascript.rhino/1.7.14.s1
[WARNING]   org.apache.tomcat/9.0.50
[WARNING]   com.servoy.eclipse.docgenerator/1.6.0
[WARNING]   org.eclipse.jface/3.27.0.v20220817-1444
```

so i would that expect that (just as with my normal product build if i don't include the jface plugin) that it would start downloading that (and fill that directory)
But for some reason it doesn't do that.. nothing is download and it still uses my local  generated,patched jface plugin.
Which is fine but i was just wondering, why does it now map on the local installed stuff and it didn't do that before?
